File: EPRO.Library.v3.5/EPRO.Library.v3.5/Types/NullableDateTime.cs
using System;
using ELibrary.Standard.Objects;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

// REM Using Operators
// REM On Objects Assignment = Automatically takes the value if same object Type[ int=int]
// REM Others must be allowed using Ctype

// REM Operator = is the Comparison Type

namespace ELibrary.Standard.Types
{



    /// <summary>
    /// A type that can hold both date and Nothing[Null] produces Date or NULL String
    /// </summary>
    /// <remarks></remarks>
    public class NullableDateTime
    {

        #region Constructors


        public NullableDateTime(DateTime DateTimeVal)
        {
            setValue(DateTimeVal);
        }

        public NullableDateTime(string DateTimeVal)
        {
            setValue(DateTimeVal);
        }

        public NullableDateTime(DBNull DateTimeVal)
        {
            __isNull = true;
        }

        public NullableDateTime(object DateTimeVal)
        {
            try
            {
                if (Information.IsNothing(DateTimeVal))
                    break;
                if (DateTimeVal is DBNull)
                    break;
                if (DateTimeVal is string)
                {
                    setValue(Conversions.ToString(DateTimeVal));
                    return;
                }

                if (DateTimeVal is DateTime)
                {
                    setValue(Conversions.ToDate(DateTimeVal));
                    return;
                }

                if (DateTimeVal is NullableDateTime)
                {
                    setValue((NullableDateTime)DateTimeVal);
                    return;
                }
            }
            catch (Exception ex)
            {
            }

            __isNull = true;
        }



        #endregion


        #region Properties


        public static readonly NullableDateTime NULL_TIME = new NullableDateTime(Constants.vbNullString);
        public static readonly NullableDateTime NOW_DATETIME = new NullableDateTime(DateTime.Now);

        /// <summary>
        /// the value it returns instead of nothing
        /// </summary>
        /// <remarks></remarks>
        public const string NULL_RETURN = "NULL";
        private DateTime ____DateTimeVal;

        public string DateValue
        {
            get
            {
                if (!isNull)
                {
                    return EDateTime.valueOf(____DateTimeVal, EDateTime.DateFormats.DateFormatsEnum.DateFormat1);
                }

                return NULL_RETURN;
            }
        }

        public DateTime DateTimeValue
        {
            get
            {
                if (!isNull)
                {
                    return ____DateTimeVal;
                }

                return default;
            }
        }

        public string TimeValue
        {
            get
            {
                if (!isNull)
                {
                    return EDateTime.time__valueOf(____DateTimeVal);
                }

                return NULL_RETURN;
            }
        }

        private bool __isNull = true;

        public bool isNull
        {
            get
            {
                return __isNull;
            }
        }

        /// <summary>
        /// returns DateTimeValue if it is not null else returns NULL string 
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public object DateTimeValueOrNULL
        {
            get
            {
                if (!isNull)
                    return DateTimeValue;
                return NULL_RETURN;
            }
        }

        /// <summary>
        /// returns DateTimeValue if it is not null else returns NULL object (NOTHING) 
        /// </summary>
        /// <value></value>
        /// <returns></returns>
        /// <remarks></remarks>
        public object DateTimeValueOrNothing
        {
            get
            {
                if (!isNull)
                    return DateTimeValue;
                return null;
            }
        }


        #endregion


        #region Operators

        public static bool operator ==(NullableDateTime ___NullableDateTime, object compareValue)
        {
            var cmpVal = new NullableDateTime(compareValue);
            if (___NullableDateTime.isNull & cmpVal.isNull)
                return true;
            if (___NullableDateTime.isNull | cmpVal.isNull)
                return false;
            return ___NullableDateTime.DateTimeValue.Equals(cmpVal.DateTimeValue);
        }

        public static bool operator !=(NullableDateTime ___NullableDateTime, object compareValue)
        {
            return !Operators.ConditionalCompareObjectEqual(___NullableDateTime, compareValue, false);
        }




        #endregion



        #region Methods


        public void setValue(string DateTimeVal)
        {
            try
            {
                if (Information.IsNothing(DateTimeVal))
                    break;
                if (string.IsNullOrEmpty(DateTimeVal))
                    break;
                ____DateTimeVal = Conversions.ToDate(DateTimeVal);
                __isNull = false;
                return;
            }
            catch (Exception ex)
            {
            }

            __isNull = true;
        }

        public void setValue(DateTime DateTimeVal)
        {
            try
            {
                if (DateTimeVal.Equals(default))
                    break;
                ____DateTimeVal = DateTimeVal;
                __isNull = false;
                return;
            }
            catch (Exception ex)
            {
            }

            __isNull = true;
        }

        private void setValue(NullableDateTime __nullableDateTime)
        {
            if (__nullableDateTime is null)
            {
                __isNull = true;
                return;
            }

            ____DateTimeVal = __nullableDateTime.DateTimeValue;
            __isNull = __nullableDateTime.isNull;
        }

        public override bool Equals(object obj)
        {
            return Operators.ConditionalCompareObjectEqual(this, obj, false);
        }

        public override string ToString()
        {
            return DateTimeValue.ToString();
        }

        #endregion



    }
}

File: EPRO.Library.v3.5/EPRO.Library.v3.5/Types/NullableDateTime.vb
Option Explicit On
Option Strict On

Imports EPRO.Library.v3._5.Objects

REM Using Operators
REM On Objects Assignment = Automatically takes the value if same object Type[ int=int]
REM Others must be allowed using Ctype

REM Operator = is the Comparison Type

Namespace Types



    ''' <summary>
    ''' A type that can hold both date and Nothing[Null] produces Date or NULL String
    ''' </summary>
    ''' <remarks></remarks>
    Public Class NullableDateTime

#Region "Constructors"


        Sub New(ByVal DateTimeVal As Date)
            Me.setValue(DateTimeVal)

        End Sub

        Sub New(ByVal DateTimeVal As String)

            Me.setValue(DateTimeVal)


        End Sub


        Sub New(ByVal DateTimeVal As DBNull)

            Me.__isNull = True
        End Sub



        Sub New(ByVal DateTimeVal As Object)

            Try

                If IsNothing(DateTimeVal) Then Exit Try

                If TypeOf DateTimeVal Is DBNull Then Exit Try

                If TypeOf DateTimeVal Is String Then Me.setValue(CStr(DateTimeVal)) : Exit Sub

                If TypeOf DateTimeVal Is Date Then Me.setValue(CDate(DateTimeVal)) : Exit Sub

                If TypeOf DateTimeVal Is NullableDateTime Then Me.setValue(CType(DateTimeVal, NullableDateTime)) : Exit Sub

            Catch ex As Exception

            End Try

            Me.__isNull = True
        End Sub



#End Region


#Region "Properties"


        Public Shared ReadOnly NULL_TIME As New NullableDateTime(vbNullString)

        Public Shared ReadOnly NOW_DATETIME As New NullableDateTime(Date.Now)

        ''' <summary>
        ''' the value it returns instead of nothing
        ''' </summary>
        ''' <remarks></remarks>
        Public Const NULL_RETURN As String = "NULL"


        Private ____DateTimeVal As Date

        Public ReadOnly Property DateValue As String
            Get
                If Not Me.isNull Then
                    Return EDateTime.valueOf(Me.____DateTimeVal, EDateTime.DateFormats.DateFormatsEnum.DateFormat1)
                End If
                Return NULL_RETURN
            End Get
        End Property

        Public ReadOnly Property DateTimeValue As Date
            Get
                If Not Me.isNull Then
                    Return Me.____DateTimeVal
                End If
                Return Nothing
            End Get
        End Property


        Public ReadOnly Property TimeValue As String
            Get
                If Not Me.isNull Then
                    Return EDateTime.time__valueOf(Me.____DateTimeVal)
                End If
                Return NULL_RETURN
            End Get
        End Property


        Private __isNull As Boolean = True
        Public ReadOnly Property isNull As Boolean
            Get
                Return Me.__isNull
            End Get
        End Property

        ''' <summary>
        ''' returns DateTimeValue if it is not null else returns NULL string 
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property DateTimeValueOrNULL As Object
            Get
                If Not Me.isNull Then Return Me.DateTimeValue
                Return NULL_RETURN
            End Get
        End Property

        ''' <summary>
        ''' returns DateTimeValue if it is not null else returns NULL object (NOTHING) 
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public ReadOnly Property DateTimeValueOrNothing As Object
            Get
                If Not Me.isNull Then Return Me.DateTimeValue
                Return Nothing
            End Get
        End Property


#End Region


#Region "Operators"

        Public Shared Operator =(ByVal ___NullableDateTime As NullableDateTime,
                                  ByVal compareValue As Object) As Boolean
            Dim cmpVal As New NullableDateTime(compareValue)
            If ___NullableDateTime.isNull And cmpVal.isNull Then Return True

            If ___NullableDateTime.isNull Or cmpVal.isNull Then Return False

            Return ___NullableDateTime.DateTimeValue.Equals(cmpVal.DateTimeValue)

        End Operator

        Public Shared Operator <>(ByVal ___NullableDateTime As NullableDateTime,
                                 ByVal compareValue As Object) As Boolean

            Return Not (___NullableDateTime = compareValue)


        End Operator




#End Region



#Region "Methods"


        Public Sub setValue(ByVal DateTimeVal As String)


            Try

                If IsNothing(DateTimeVal) Then Exit Try
                If DateTimeVal = vbNullString Then Exit Try


                Me.____DateTimeVal = CDate(DateTimeVal)
                Me.__isNull = False
                Exit Sub

            Catch ex As Exception

            End Try
            Me.__isNull = True

        End Sub

        Public Sub setValue(ByVal DateTimeVal As Date)
            Try

                If DateTimeVal.Equals(Nothing) Then Exit Try


                Me.____DateTimeVal = DateTimeVal
                Me.__isNull = False
                Exit Sub

            Catch ex As Exception

            End Try
            Me.__isNull = True

        End Sub

        Private Sub setValue(ByVal __nullableDateTime As NullableDateTime)
            If __nullableDateTime Is Nothing Then Me.__isNull = True : Exit Sub
            Me.____DateTimeVal = __nullableDateTime.DateTimeValue
            Me.__isNull = __nullableDateTime.isNull

        End Sub

        Public Overrides Function Equals(ByVal obj As Object) As Boolean
            Return (Me = obj)
        End Function

        Public Overrides Function ToString() As String
            Return Me.DateTimeValue.ToString()
        End Function

#End Region



    End Class


End Namespace